parsl/executors/high_throughput/mpi_resource_management.py

```diff
@@ -0,0 +1,222 @@
+from parsl.serialize import (pack_res_spec_apply_message,
+                             unpack_res_spec_apply_message)
+
+logger = logging.getLogger(__name__)
+
+
+class Scheduler(Enum):
+    Unknown = 0
+    Slurm = 1
+    PBS = 2
+    Cobalt = 3
+
+
+def get_slurm_hosts_list() -> List[str]:
+    """Get list of slurm hosts from scontrol"""
+    cmd = "scontrol show hostname $SLURM_NODELIST"
+    b_output = subprocess.check_output(
+        cmd, stderr=subprocess.STDOUT, shell=True
+    )  # bytes
+    output = b_output.decode().strip().split()
+    return output
+
+
+def get_pbs_hosts_list() -> List[str]:
+    """Get list of PBS hosts from envvar: PBS_NODEFILE"""
+    nodefile_name = os.environ["PBS_NODEFILE"]
+    with open(nodefile_name) as f:
+        return [line.strip() for line in f.readlines()]
+
+
+def get_cobalt_hosts_list() -> List[str]:
+    """Get list of COBALT hosts from envvar: COBALT_NODEFILE"""
+    nodefile_name = os.environ["COBALT_NODEFILE"]
+    with open(nodefile_name) as f:
+        return [line.strip() for line in f.readlines()]
+
+
+def get_nodes_in_batchjob(scheduler: Scheduler) -> List[str]:
+    """Get nodelist from all supported schedulers"""
+    nodelist = []
+    if scheduler == Scheduler.Slurm:
+        nodelist = get_slurm_hosts_list()
+    elif scheduler == Scheduler.PBS:
+        nodelist = get_pbs_hosts_list()
+    elif scheduler == Scheduler.Cobalt:
+        nodelist = get_cobalt_hosts_list()
+    return nodelist
+
+
+def identify_scheduler() -> Scheduler:
+    """Use envvars to determine batch scheduler"""
+    if os.environ.get("SLURM_NODELIST"):
+        return Scheduler.Slurm
+    elif os.environ.get("PBS_NODEFILE"):
+        return Scheduler.PBS
+    elif os.environ.get("COBALT_NODEFILE"):
+        return Scheduler.Cobalt
+    else:
+        return Scheduler.Unknown
+
+
+class MPIResourceUnavailable(Exception):
+    """Raised if there are no free resources available for an MPI request"""
+
+    def __init__(self, requested: int, available: int):
+        self.requested = requested
+        self.available = available
+
+    def __str__(self):
+        return f"MPIResourceUnavailable(requested={self.requested} available={self.available}"
+
+
+class TaskScheduler:
+    """Default TaskScheduler that does no taskscheduling
+
+    This class simply acts as an abstraction over the task_q and result_q
+    that can be extended to implement more complex task scheduling logic
+    """
+    def __init__(
+        self,
+        pending_task_q: multiprocessing.Queue,
+        pending_result_q: multiprocessing.Queue,
+    ):
+        self.pending_task_q = pending_task_q
+        self.pending_result_q = pending_result_q
+
+    def put_task(self, task) -> None:
+        return self.pending_task_q.put(task)
+
+    def get_result(self, block: bool, timeout: float):
+        return self.pending_result_q.get(block, timeout=timeout)
+
+    def put_result(self, result) -> None:
```

Comment:
why is put_result, the far end of the communication abstracted here but not get_task?

Reply:
Tasks when scheduled are placed into the `pending_task_q` which is a`mp.Queue` from which the workers will get the task. The task scheduler abstraction didn't make much sense from the worker side, so I didn't implement the worker side primitive.

Reply:
put_result probably doesn't make sense to abstract on the worker side then?